lib/gc/Transforms/StaticMemoryPlanning.cpp

```diff
@@ -0,0 +1,656 @@
+        // itr points to the last element
+        --itr;
+        // if not found, this means that all free chunk is smaller than
+        // aligned size, switch to the largest chunk
+        target = itr->second;
+        extendAlloc(target, aligned);
+        return target;
+      } else {
+        MemoryChunk *rhs;
+        return splitAlloc(itr->second, aligned, rhs);
+      }
+    }
+  }
+
+  void dealloc(MemoryChunk *chk) {
+    tick++;
+    chk->lastFreedTick = tick;
+    addChunkToFreeList(chk);
+    auto itrInCurChunks = std::find(curChunks.begin(), curChunks.end(), chk);
+    assert(itrInCurChunks != curChunks.end());
+    // merge left and right if they are free
+    std::vector<MemoryChunk *>::iterator toMergeStart = itrInCurChunks;
+    std::vector<MemoryChunk *>::iterator toMergeEnd = itrInCurChunks + 1;
+    // look left to see any one we can merge with
+    for (auto itr = itrInCurChunks;; --itr) {
+      if ((*itr)->isfree) {
+        toMergeStart = itr;
+      } else {
+        break;
+      }
+      if (itr == curChunks.begin()) {
+        break;
+      }
+    }
+    // look right to see any one we can merge with
+    for (auto itr = itrInCurChunks + 1; itr != curChunks.end(); ++itr) {
+      if ((*itr)->isfree) {
+        toMergeEnd = itr + 1;
+      } else {
+        break;
+      }
+    }
+    if (toMergeEnd - toMergeStart > 1) {
+      // do merge
+      chunks.emplace_back(std::make_unique<MergedChunk>(
+          std::vector<MemoryChunk *>(toMergeStart, toMergeEnd)));
+
+      // remove merged chunks from free list and cur_chunk list
+      for (auto itr = toMergeStart; itr != toMergeEnd; ++itr) {
+        auto chunk = *itr;
+        removeChunkFromFreeList(chunk);
+      }
+      // add merged chunk to cur_chunks and free_chunks_by_size
+      *toMergeStart = chunks.back().get();
+      chunks.back()->lastFreedTick = tick;
+      addChunkToFreeList(chunks.back().get());
+      curChunks.erase(toMergeStart + 1, toMergeEnd);
+    }
+    // else, no chunks are merged, do nothing
+  }
+
+  void dealloc(uintptr_t bufferid) {
+    auto alocitr = allocations.find(bufferid);
+    assert(alocitr != allocations.end() &&
+           "Cannot find buffer id in allocations");
+    auto itr = curAllocations.find(bufferid);
+    if (itr != curAllocations.end()) {
+      itr->second->isInplaceSplitRemainder = false;
+      dealloc(itr->second);
+      curAllocations.erase(itr);
+    }
+  }
+
+  std::string toString() const {
+    std::stringstream ss;
+    ss << "total size " << currentAllocSize << " ";
+    size_t cur_offset = 0;
+    for (auto buf : curChunks) {
+      ss << "| " << cur_offset << ',' << buf->size << ',' << buf->isfree << " ";
+      cur_offset += buf->size;
+    }
+    return ss.str();
+  }
+};
+
+size_t scheduleMemoryAllocations(
+    const Traces &traces, std::size_t alignment, bool hotFirst,
+    const InplaceInfoMap &inplaceMap,
+    std::unordered_map<uintptr_t, std::size_t> &outSchedule,
```

Comment:
One open here: is there any possible to estimate allocation size during fusion stage, just like what we do in v1?

Reply:
Yes. The internals of the pass is already described at the doc. You can skip the IR mutation stage and get the mem scheduling results.

Reply:
After offline sync, this pass expects `memref` rather than `tensor` dialect. So, it needs more changes to be applied during fusion stage.